### PR TITLE
return completion item kind as function

### PR DESCRIPTION
**Pull request by xiantang**
image

to 

image

### CODE REVIEW

File: lsp/lsp.jq [hunk 1]
```diff
@@ -675,7 +677,8 @@ def handle($state):
             | if .args and (.args | length) > 0 then
                 { label: "\(.str)/\(.args | length)",
                   insertText: "\(.str)($1)",
-                  insertTextFormat: TextFormatSnippet
+                  insertTextFormat: TextFormatSnippet,
+                  kind: CompletionItemKindFunction,
```
Comment (xiantang) [inline lsp/lsp.jq:681]: maybe we should add document also

File: lsp/lsp.jq [hunk 2]
```diff
@@ -675,7 +677,9 @@ def handle($state):
...
+                  documentation: docs[env_func_name],
```
Comment (wader) [inline lsp/lsp.jq:682]: Nice, i will check how it looks in vscode. Possibly show full function signature with arg name, similar to how textDocument/hover respond?

Comment (xiantang) [reply lsp/lsp.jq:682]: idk how to show full function in signature, could u help me copy it from hover.

File: lsp/lsp.jq [hunk 3]
```diff
@@ -675,7 +677,9 @@ def handle($state):
             | if .args and (.args | length) > 0 then
                 { label: "\(.str)/\(.args | length)",
                   insertText: "\(.str)($1)",
-                  insertTextFormat: TextFormatSnippet
+                  insertTextFormat: TextFormatSnippet,
+                  kind: CompletionItemKindFunction,
```
Comment (wader) [inline lsp/lsp.jq:681]: Maybe should be `CompletionItemKindVariable` (6) if it's a binding?

File: lsp/lsp.jq (hunk 2 above)
Comment (wader) [reply lsp/lsp.jq:682]: sure will have a look

Comment (xiantang) [reply lsp/lsp.jq:682]: image


seems have small problem
```
                { label: "\(.str)/\(.args | length)",
                  insertText: "\(.str)($1)",
                  insertTextFormat: TextFormatSnippet,
                  kind: CompletionItemKindFunction,
                  documentation: (docs[env_func_name] as $doc |[ "
                  ```jq"
                  , "def \(env_func_signature):"
                  , "```"
                  , if $doc then $doc else empty end
                    ]
                  | join("\n")
                )}

```

Comment (wader) [reply lsp/lsp.jq]: Ok maybe have to skip markdown. Does neovim use/show hover documentation somehow?

File: lsp/lsp.jq (hunk 3 above)
Comment (xiantang) [reply lsp/lsp.jq:681]: yes we should have condition there

File: lsp/lsp.jq (hunk 2 above)
Comment (xiantang) [reply lsp/lsp.jq]: image
show hover like this. gopls

Comment (xiantang) [reply lsp/lsp.jq]: maybe need use this https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#markupContentInnerDefinition

Comment (xiantang) [reply lsp/lsp.jq]: image

done

Comment (wader) [reply lsp/lsp.jq]: Screenshot 2022-12-20 at 10 17 19

vscode seems to not understand it's markdown, probably some small thing, looking into it

We should also move the documentation out of `if .args and (.args | length) > then` so it's done for funtions with 0 args

Comment (wader) [reply lsp/lsp.jq]: It was the whitespace before "```jq"

**Comment by wader** [comment]
Did some refactor and fixe markdown in vscode, maybe something like this:
```patch
diff --git a/lsp/lsp.jq b/lsp/lsp.jq
index 94494f5..48efc8a 100644
--- a/lsp/lsp.jq
+++ b/lsp/lsp.jq
@@ -97,6 +97,8 @@ def CompletionItemKindFunction: 3;
 
 def TextDocumentSyncFull: 1;
 
+def MarkupKindsMarkdown: "markdown";
+
 def SymbolKindFile: 1;
 def SymbolKindModule: 2;
 def SymbolKindNamespace: 3;
@@ -199,6 +201,16 @@ def env_func_signature:
   | join("")
   );
 
+def env_func_markdown:
+  ( docs[env_func_name] as $doc
+  | [ "```jq"
+    , "def \(env_func_signature):"
+    , "```"
+    , if $doc then $doc else empty end
+    ]
+  | join("\n")
+  );
+
 def query_walk($uri; $start_env; f):
   def _t($start_env):
     def _pattern_env:
@@ -674,26 +686,20 @@ def handle($state):
             | env_iter_entries
             | .value
             | select(.str | startswith($prefix))
+            | . as $func
             | if .args and (.args | length) > 0 then
                 { label: "\(.str)/\(.args | length)",
                   insertText: "\(.str)($1)",
                   insertTextFormat: TextFormatSnippet,
                   kind: CompletionItemKindFunction,
-                  documentation: {
-                  value :(docs[env_func_name] as $doc |[ "
-                  ```jq"
-                  , "def \(env_func_signature):"
-                  , "```"
-                  , if $doc then $doc else empty end
-                    ]
-                  | join("\n")
-                  ),
-                  kind:"markdown",
-                  },
                 }
               else
                 {label: .str}
               end
+            | .documentation =
+                { value: ($func | env_func_markdown),
+                  kind: MarkupKindsMarkdown
+                }
             ]
           )
         )
@@ -754,14 +760,7 @@ def handle($state):
           )
         | docs[env_func_name] as $doc
         | result({
-            contents:
-              ( [ "```jq"
-                , "def \(env_func_signature):"
-                , "```"
-                , if $doc then $doc else empty end
-                ]
-              | join("\n")
-              )
+            contents: env_func_markdown
           })
         )
       else

```

File: lsp/lsp.jq (hunk 2 above)
Comment (xiantang) [reply lsp/lsp.jq]: yes i've updated, plz try again

Comment (wader) [reply lsp/lsp.jq]: Screenshot 2022-12-20 at 10 37 35
👍

Comment (wader) [reply lsp/lsp.jq]: and conditional for variable like this maybe:
```jq
```

**Comment by wader** [comment]
Sorry for all the small snippet and comments, but this will be nice addition, thanks!

**Comment by wader** [comment]
Also sorry for horrible dev environment and no tests :( jq-lsp was kind of a quick hack that turned out to work much better than i expected :)

**Comment by xiantang** [comment]
> Sorry for all the small snippet and comments, but this will be nice addition, thanks!

I have put added your snippet.

**Comment by xiantang** [comment]
> Also sorry for horrible dev environment and no tests :( jq-lsp was kind of a quick hack that turned out to work much better than i expected :)

its useful

**Comment by wader** [comment]
LGTM, ok to merge?

If we want to add more things maybe better to do in separate PR, this one is already growing a bit large

**Comment by xiantang** [comment]
> LGTM, ok to merge?
> 
> If we want to add more things maybe better to do in separate PR, this one is already growing a bit large

yes. plz merge. this pr is a little large now.

**Pull request merged**